### ISSUE TITLE
Meta-ticket: Cython 3.0.0 (released July 2023)

### ISSUE THREAD

**Issue by mkoeppe** [sagemath/sage#29863]
<div id="comment:0"></div>

This ticket is for experimenting with an update to Cython 3, which entered the beta stage on 2023-02-25 after several years in alpha. (No release schedule has been published.)
- https://mail.python.org/pipermail/cython-devel/2019-November/thread.html (2019 - no response)
- https://groups.google.com/g/cython-users/c/WTHmGJ44aB0 (2022-01 "no concrete time plan")
- https://github.com/cython/cython/issues/4022 (still no road map)

https://github.com/cython/cython/blob/master/CHANGES.rst

- #34257 Add missing `cdef` method declarations (for Cython 3)
- #26254 Use Cython directive `binding=True` to get signatures in help for cythonized built-in methods
- #32832 `cython`: Eliminate use of custom patches
- #30511 Type hints (`__annotations__`) and coercion/categories/Cython

See also:
- #34221 / https://github.com/cython/cython/pull/4918 (Backport to 0.29.x: Add PEP420 namespace support)

Depends on #26254
Depends on #32832
Depends on #34257

CC:  @kwankyu @williamstein @videlec @tscrim @tobiasdiez

Component: **packages: standard**

Keywords: **upgrade**

Branch/Commit: **[u/mkoeppe/meta_ticket__cython_3_0_0__unreleased_](https://github.com/sagemath/sagetrac-mirror/tree/u/mkoeppe/meta_ticket__cython_3_0_0__unreleased_) @ [`6c338f8`](https://github.com/sagemath/sagetrac-mirror/commit/6c338f8050c77f97d8b597d855b41547e7c06c22)**

_Issue created by migration from https://trac.sagemath.org/ticket/29863_

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1,5 +1,9 @@
-This ticket is for experimenting with an update to Cython 3, currently at 3.0.0 alpha 5.
+This ticket is for experimenting with an update to Cython 3, which is at 3.0.0 alpha 10 as of 2022-07. No release schedule has been published (https://mail.python.org/pipermail/cython-devel/2019-November/thread.html)
 
 https://github.com/cython/cython/blob/master/CHANGES.rst
 
+- #26254 Use Cython directive binding=True to get signatures in help for cythonized built-in methods
+- #30511 Type hints (__annotations__) and coercion/categories/Cython
+- https://github.com/cython/cython/pull/4918 (Backport to 0.29.x: Add PEP420 namespace support)
 
+
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -2,8 +2,8 @@
 
 https://github.com/cython/cython/blob/master/CHANGES.rst
 
-- #26254 Use Cython directive binding=True to get signatures in help for cythonized built-in methods
-- #30511 Type hints (__annotations__) and coercion/categories/Cython
+- #26254 Use Cython directive `binding=True` to get signatures in help for cythonized built-in methods
+- #30511 Type hints (`__annotations__`) and coercion/categories/Cython
 - https://github.com/cython/cython/pull/4918 (Backport to 0.29.x: Add PEP420 namespace support)
 
 
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1,4 +1,7 @@
-This ticket is for experimenting with an update to Cython 3, which is at 3.0.0 alpha 10 as of 2022-07. No release schedule has been published (https://mail.python.org/pipermail/cython-devel/2019-November/thread.html)
+This ticket is for experimenting with an update to Cython 3, which is at 3.0.0 alpha 10 as of 2022-07. No release schedule has been published.
+- https://mail.python.org/pipermail/cython-devel/2019-November/thread.html (2019 - no response)
+- https://groups.google.com/g/cython-users/c/WTHmGJ44aB0 (2022-01 "no concrete time plan")
+- https://github.com/cython/cython/issues/4022
 
 https://github.com/cython/cython/blob/master/CHANGES.rst
 
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -1,7 +1,7 @@
 This ticket is for experimenting with an update to Cython 3, which is at 3.0.0 alpha 10 as of 2022-07. No release schedule has been published.
 - https://mail.python.org/pipermail/cython-devel/2019-November/thread.html (2019 - no response)
 - https://groups.google.com/g/cython-users/c/WTHmGJ44aB0 (2022-01 "no concrete time plan")
-- https://github.com/cython/cython/issues/4022
+- https://github.com/cython/cython/issues/4022 (still no road map)
 
 https://github.com/cython/cython/blob/master/CHANGES.rst
 
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -6,7 +6,10 @@
 https://github.com/cython/cython/blob/master/CHANGES.rst
 
 - #26254 Use Cython directive `binding=True` to get signatures in help for cythonized built-in methods
+- #32832 `cython`: Eliminate use of custom patches
 - #30511 Type hints (`__annotations__`) and coercion/categories/Cython
+
+See also:
 - https://github.com/cython/cython/pull/4918 (Backport to 0.29.x: Add PEP420 namespace support)
 
 
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -10,6 +10,6 @@
 - #30511 Type hints (`__annotations__`) and coercion/categories/Cython
 
 See also:
-- https://github.com/cython/cython/pull/4918 (Backport to 0.29.x: Add PEP420 namespace support)
+- #34221 / https://github.com/cython/cython/pull/4918 (Backport to 0.29.x: Add PEP420 namespace support)
 
 
``````

**Comment by mkoeppe** [comment]
Description changed:
``````diff
--- 
+++ 
@@ -5,6 +5,7 @@
 
 https://github.com/cython/cython/blob/master/CHANGES.rst
 
+- #34257 Add missing `cdef` method declarations (for Cython 3)
 - #26254 Use Cython directive `binding=True` to get signatures in help for cythonized built-in methods
 - #32832 `cython`: Eliminate use of custom patches
 - #30511 Type hints (`__annotations__`) and coercion/categories/Cython
``````

**Comment by tornaria** [comment]
Now that 3.0.0 is released, could we have this rebased and in a PR, please?

**Comment by tornaria** [comment]
Related: https://github.com/cython/cython/issues/5552

**Comment by tornaria** [comment]
Cf: https://github.com/tornaria/sage/tree/cython3-legacy, I've got as far as being able to cythonize everything, but some C files still fail to compile.

First five commits are `git cherry-pick f36275a7869 36ec4930832 1f9ee80cded 034094c849e 9d89bdf776` from the sagetrac mirror on top of 10.1.beta7.

Note that https://github.com/cython/cython/issues/5552 means cythonize fails if python package lxml is available. Can use `SAGE_DEBUG=no` to avoid the issue which only happens with debug enabled.

Also: https://github.com/void-linux/void-packages/pull/45086 includes patches for cysignals, cypari2, fpylll and pplpy that might be useful.

**Comment by tornaria** [comment]
After adding a commit fixing inheritance in `lcalc_Lfunction` (see https://github.com/sagemath/sage/compare/10.1.beta7...tornaria:sage:d0ad68ed9d4) I've got the whole of sagelib to compile.

Now it fails at `import sage.structure.element` as follows:
```
$ PYTHONPATH=pkgs/sagemath-standard/build/lib.linux-x86_64-cpython-311 python
Python 3.11.4 (main, Jun  8 2023, 02:02:15) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sage.structure.element
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../pkgs/sagemath-standard/build/lib.linux-x86_64-cpython-311/sage/structure/__init__.py", line 2, in <module>
    import sage.structure.element
  File "sage/structure/element.pyx", line 1, in init sage.structure.element (build/cythonized/sage/structure/element.c:44218)
TypeError: PyMethodDescr_CallSelf requires a method without arguments
```

*Please note:* I'll probably keep rebasing and/or changing commits in my branch https://github.com/tornaria/sage/tree/cython3-legacy so don't take them as public commits unless you are willing to rebase.

**Comment by tornaria** [comment]
> Branch pushed to git repo; I updated commit sha1. New commits:
> [`fd2e122`](https://github.com/sagemath/sagetrac-mirror/commit/fd2e122cc0b8d7cba07efdcfab3e54436f26aec1)	`src/sage/libs/lcalc/lcalc_Lfunction.pxd: Add missing cdef method declarations`

This is incorrect and it stems from a fundamental change in cython 3 that will probably bite us all over the place (and this one cannot be switched off afaict).

I better link to the relevant section of the migration guide: https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#class-private-name-mangling

This means sagelib has been (mis)using `__name` members (methods and attributes) that (a) in python are class-private via name mangling (b) in cython before 3.0.0 they are inherited due to mangling not implemented (c) in cython 3.0.0 they are class-private via the exact same name mangling as in python.

So the way to fix all these "missing cdef ..." errors is *not* by adding the missing methods but renaming the members to use a single underscore which is the python way to declare a "private" member that is still inherited by subclasses (think "protected" in c++).
